### ISSUE TITLE
Global Disable or Hide Keyboard in TOC jump mode only

### ISSUE THREAD

**Issue by boulderob** [gsantner/markor#1325]
#### General information

* **App version:** 

v2.6.0(122)

* **System:** 

Android 10 (galaxy tab A7 SM-T500)

#### Description

On my main computer, I regularly use and maintain a library of **LONG md docs** that i insert H{1..n} md headers into (ie md headers as pounds signs) to create a navigable md TOC in each md document.   This allows me to ignore major portions of the documents in that library and once loaded in a md program / editor like vim, to jump to only the portions (headers) i want to quickly and easily.  The md TOC is imperative to my workflow.  Without the TOC, i would have to switch between and then navigate with long documents and it becomes unusable.

I recently purchased a samsung galaxy A7 tablet so that i could sync this lib of **LONG** TOC md docs to it and take it with me on the road.  In that sense, Markor's role as an editor is secondary to me.  I want Markor to mostly be a smart md reader with good TOC features.  It's almost there but my workflow could be improved with a number of configuration enhancements.  I'm not sure if there are some fixes already present.  I've scoured the app, the forum and the web trying to look for fixes on my own.  

I'm coming from the iOS world so forgive me if I'm missing something easy here but I've put a lot of effort trying to solve this on my own before posting here.  

Thoughts welcome.

# KEYBOARD DISABLE

The markor editor window TOC that exists at the top of each opened file works almost exactly how i want it.  It transforms my doc md headers into TOC jump links.   That allows me to easily click on the top of the window, pull down the TOC and jump to anywhere in the doc **regardless of where i'm at in the document which is a big deal in long documents**!  love it!

But an annoyance for my particular workflow in the editor is that every time i jump using the TOC dropdown menu, since it naturally moves the cursor to the appropriate place in the document, the default samsung keyboard pops out and clutters the editor.  This means that every time i jump i have to manually "hide" the keyboard down and off the screen.  **It becomes really tiresome because i jump around a lot in these long MD docs**.

I'm sure that's probably normal editor behavior for most use cases but **I'm mostly using the Markor editor as a TOC md jump reader**!  

I tried disabling the samsung keyboard via the system and it won't let me (ie it's always dimmed so I can't turn it off).  Internet search tells me that it's not possible b/c then you wouldn't have a keyboard to login and you'd be locked out of your device and have to factory reset it. 

## POSSIBILITY ONE:

Perhaps this is really simple and I just don't know how to do this in either Android or Markor but **being able to disable the keyboard globally/outright within the app (ie at the app level) with a Markor interface icon/widget toggle checkbox or settings config option** would be perfect for my use case.  Is there anyway to do what i want already or would this need to be an app enhancement?  If the latter would you consider it?

## POSSIBILITY TWO

The alternate and perhaps better method would be to not disable keyboard at the global app level but configure it to "hide" / not pop out only when Markor jumps from the TOC (ie moves the cursor).   Is a configuration option to en|disable this type of functionality possible.  

I'd personally love this feature because **i jump a lot in non-editing contexts** and i have to keep manually "hiding" the keyboard all the time.  It's counterproductive to an efficient workflow.

# ALTERNATIVE SOLUTION : USE PREVIEW vs EDIT MODE

Your first reaction might be to tell me to just use the Markor Preview instead of Edit mode and then make Preview the default.

The problem with that is that

1. first and foremost, preview mode TOC to my knowledge does **not** work like edit mode TOC. 
 
 By that i mean that **edit mode TOC is always ready and armed from a drop-down menu at the top of the edit mode doc window** which means i can be 70 pages down into my long document and simply click the top of the "app doc bar" TOC dropdown and jump where i want.  

**In preview mode, the TOC becomes INLINE HTML and always exists at the top of the preview document / html file itself.  This requires you physically scroll up 70 or 200 pages or whatever before you can even access it and that is unusable for any kind of long document navigation as is**.   

Basically the edit mode TOC is the right way to implement this for long documents so that it's practical and usable.  Without an extra and separate pane for the TOC in preview mode that allows you to show/hide it, it's more gimmick then real functionality for long documents.

2. although my use case is primarily Markor as a reader, i actually would love to highlight (bold, italic, underline) phrases from time to time.  So navigating the editor using the TOC primarily as a reader with ability to manually show/hide the keyboard out when i want it would be great.

thx

**Comment by aubreyz** [comment]
The summary of all that is: It would be nice if, within edit mode, there was some sort of dropdown menu permanently at the top to jump to a particular header (without activating the keyboard). I agree that this would be a nice feature if possible. Perhaps a simpler feature would be a restricted search button to jump to the next or previous header (or perhaps next or previous header of a particular level).

**Comment by boulderob** [comment]
@aubreyz that is correct and very succinct. :)  i thought it would be better to be specific / clear since activity seems a bit sporadic here.  this allows others to move ahead without lots of back and forth once they read it and or me having to check back and reply for clarifications which is often the case.  in good faith, i've hopefully thought through the alternatives and put some effort in on my end which might help move things forward.

"editorial" md editor app works as i describe on iOS.  it's been out for many years now but isn't actively developed much and not at all for android.

**Comment by gsantner** [comment]
Sorry but I couldn't get time to read through this all. For sure, there is no interest to add an additional settings option for this quite specific request. Rather the "improve the default way how it works" is the way to go.

**Comment by harshad1** [comment]
@boulderob Can't you just use view mode?

**Comment by boulderob** [comment]
> @boulderob Can't you just use view mode?

@harshad1 I explained thoroughly in my original post why that is suboptimal especially for very long documents.  Check it out!

**Comment by gsantner** [comment]
Long press the top-bar button, jumps to top/bottom of the document. Works in both edit & view mode in same way.

**Comment by harshad1** [comment]
> Long press the top-bar button, jumps to top/bottom of the document. Works in both edit & view mode in same way.

To add to this, you can grab the scroll bar and drag to scroll fast.

**Comment by boulderob** [comment]
Perhaps we have a disconnect.  If I have a file that's 200 pages long with 200 headers and i'm on 150

- in edit mode i just tap on the TOC and i immediately have my TOC.  One click and I'm right there to go to my next move. BUT, if i cancel, my cursor is still where it is on 150.  I never have to actually change my focus / context until i click a new header in the TOC.  in effect, the TOC is an overlay in edit mode.  

- in preview mode, i didn't realize i could long-press to go to top (thanks for the tip!). But in preview mode the TOC is *not* an overlay.  It's inline uneditable text.  When you scroll from page 150 you are jumping ie changing the position of the cursor to view the TOC.   That's quite a heavy move when an overlay which already exists in edit mode works.  With the jump to TOC in preview mode, If i change my mind, or want to cxl I'm screwed.  I'd have to remember where i last was to go back probably scrolling.   

- preview mode also doesn't allow me to edit if and when i want to.

- of course I can scroll and often do.  But, i go to great pains to inject headers in key portions of long documents so i don't have to scroll to find them.  That's the beauty of a TOC :)  Otherwise I'd be scrolling around all day looking for things that are really hard to find that i've already taken the time to mark for later reference.   If you've ever worked with a library of long documents that you use on a regular basis you will understand the benefit of the TOC vs scrolling.  Create a 200 page document and try scrolling around.  You waste lots of brain energy scrolling around looking for things.  

make sense?

**Comment by boulderob** [comment]
I'll add one more good reason why disabling the auto keyboard visibility on jump from TOC in edit mode is a good feature. 

When you jump to a header with the intent to edit, more often than not you probably do *not* want to edit that exact header line.  You want the jump to get you to the relevant section and then you're going to scroll a bit to find a paragraph or sentence in that section.  Sometimes it's immediately after the header.  But it could easily be after the last sentence in the first paragraph,  the first sentence in the last paragraph or any other sentence/paragraph combination in between.

In such cases, the autovisibility of the keyboard just clogs the screen and prevents you from locating the thing you want faster and easier.  Since you're already likely to drag / scroll a bit to fine-tune the edit point, the subsequent tap to set the cursor to that *specific* location would ideally then trigger the keyboard visibility.   You have to make that second tap anyway to set the cursor so for all other cases except wanting to edit exactly on the header line, you're not eliminating any taps the user has to make. 

So the only change to make this work is simply configuring TOC jump to not autoshow keyboard visibility.  Keyboard visibility is already the default for tapping to set the cursor so it will just happen when the user makes the first tap after the TOC jump.

As i stated earlier in the discussion, this is the behavior in an iOS app that i've used in the past when using it's TOC in edit mode and it really makes a lot of sense.   It's even more useful on a phone with small screen real estate where the virtual keyboard eats up most of the screen and obscures the text preventing your from scrolling to make that second tap

**Comment by harshad1** [comment]
Right now we use `setSelection` to scroll to a location ([here](https://github.com/gsantner/markor/blob/592181e695c2932c90b12ae2bc969116545564ac/app/src/main/java/net/gsantner/markor/format/markdown/MarkdownTextActions.java#L169))

If you're able to figure out how to scroll to a character index without causing android to open the keyboard, I'll make a pr

May 6, 2021 04:50:21 boulderob ***@***.***>:

**Comment by boulderob** [comment]
I've never programmed / tested android nor have I ever set up any kind of android build env.   The link to the contribution guide on the website is broken so I'd be starting from scratch.   My free time is severely limited at the moment for that amount of investment.  Would love to take it on but realistically it's not gonna happen any time soon.  I would've forked and gone about it already otherwise.  I do appreciate the link to the source file plus the info on the call to trigger the keyboard though.  It's a start.

**Comment by harshad1** [comment]
What I meant was that if you're able to figure out how to scroll without causing the keyboard to pop up (eg via a stackoverflow answer or something), _I'd_ be happy to create a pr.


May 6, 2021 16:54:19 boulderob ***@***.***>:

**Comment by boulderob** [comment]
@harshad1 Your link looks like a pretty good lead assuming the solution for android 4 is still relevant in current versions which i think are 10 going on 11(?).

I had searched on my own a few days ago and then more recently using 'setSelection' hint to try and narrow it down by myself.  I think i actually even came upon the link you just sent at one point by way of another posting about setting return to true to solve a related problem.  That post suggestion was going back 10 years ago though (which is a long time in development years)  but wasn't working anymore in the link you sent which itself goes back to 2014 and is referencing android 4.0.   Basically i got a lot of dated hits where confidence in relevancy was low having little connection with the present state of android api, so i didn't pass them on.  

The link you sent appears to have two working solutions for up to android 4.0 based on the version and requires two changes to the EditText object (the InputType and setting Boolean True via different object calls).   Perhaps it still works.  Perhaps you still need to add another else condition for some later version(s)?

If this still works, by default this seems to globally disable show keyboard.   But at some point you're going to need to show it again, so you have to account for that on the tap after the initial jump from the TOC and or decide when and how to change the state of the EditText object based upon the call to TOC.  

However, I don't have enough experience with the android api, jargon or editing text to know how all the pieces fit together enough to be driving a PR that you haven't to my knowledge created yet if that makes sense.  As worded, It sounds like you're waiting on me 100% to present solutions before even opening a PR yes?   

Thanks for looking into it in as much as you can regardless.
